### PR TITLE
React to Caching api changes

### PR DESCRIPTION
**Pull request by kichalla**
@Tratcher @pranavkm

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs [hunk 1]
```diff
@@ -43,22 +43,24 @@ public DefaultCodeTreeCache(IFileProvider fileProvider)
         public CodeTree GetOrAdd([NotNull] string pagePath,
                                  [NotNull] Func<IFileInfo, CodeTree> getCodeTree)
         {
-            return _codeTreeCache.GetOrSet(pagePath, getCodeTree, OnCacheMiss);
-        }
+            CodeTree codeTree;
+            if(!_codeTreeCache.TryGetValue(pagePath, out codeTree))
+            {
+                // GetOrAdd is invoked for each _ViewStart that might potentially exist in the path.
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs]: There's no GetOrAdd tho

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs [hunk 2]
```diff
@@ -184,9 +185,10 @@ public class CompilerCache : ICompilerCache
             var compilationResult = compile(file).EnsureSuccessful();
 
             // Concurrent addition to MemoryCache with the same key result in safe race.
-            var cacheEntry = _cache.Set(normalizedPath,
-                                        new CompilerCacheEntry(file, compilationResult.CompiledType),
-                                        PopulateCacheSetContext);
+            var compilerCacheEntry = new CompilerCacheEntry(file, compilationResult.CompiledType);
+            var cacheEntry = _cache.Set<CompilerCacheEntry>(normalizedPath,
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs]: No #YOLOBreaks. Arguments all on their own line, or all on the same.

File: test/WebSites/MvcTagHelpersWebSite/Components/ProductsViewComponent.cs [hunk 3]
```diff
@@ -16,7 +16,11 @@ public IViewComponentResult Invoke(string category)
         {
             IExpirationTrigger trigger;
             var products = ProductsService.GetProducts(category, out trigger);
-            EntryLinkHelpers.ContextLink.AddExpirationTriggers(new[] { trigger });
+
+            if (EntryLinkHelpers.ContextLink != null)
```
Comment (kichalla) [inline test/WebSites/MvcTagHelpersWebSite/Components/ProductsViewComponent.cs]: @Tratcher : Need to discuss with you about `EntryLinkHelpers`...the api review required this to be `internal`, but probably we should keep it `public` as I need to get access to it for this scenario...

File: src/Microsoft.AspNet.Mvc.TagHelpers/CacheTagHelper.cs [hunk 4]
```diff
@@ -120,6 +120,7 @@ public class CacheTagHelper : TagHelper
         /// <inheritdoc />
         public override async Task ProcessAsync(TagHelperContext context, TagHelperOutput output)
         {
+            //Enabled = false;
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.TagHelpers/CacheTagHelper.cs]: Remove?

File: src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs [hunk 5]
```diff
@@ -43,22 +43,24 @@ public DefaultCodeTreeCache(IFileProvider fileProvider)
         public CodeTree GetOrAdd([NotNull] string pagePath,
                                  [NotNull] Func<IFileInfo, CodeTree> getCodeTree)
         {
-            return _codeTreeCache.GetOrSet(pagePath, getCodeTree, OnCacheMiss);
-        }
+            CodeTree codeTree;
+            if(!_codeTreeCache.TryGetValue(pagePath, out codeTree))
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs]: Formatting.

File: src/Microsoft.AspNet.Mvc.TagHelpers/CacheTagHelper.cs [hunk 6]
```diff
@@ -128,17 +129,13 @@ public class CacheTagHelper : TagHelper
                 {
                     // Create an EntryLink and flow it so that it is accessible via the ambient
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.TagHelpers/CacheTagHelper.cs]: Comments need to be updated.

File: test/WebSites/MvcTagHelpersWebSite/Components/ProductsViewComponent.cs (hunk 3 above)
Comment (pranavkm) [inline test/WebSites/MvcTagHelpersWebSite/Components/ProductsViewComponent.cs]: Shouldn't it be available hanging off of `MemoryCache`? `MemoryCache.CreateLinkingScope()` -> `MemoryCache.CurrentLinkingScope`?

File: src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorPreCompiler.cs [hunk 7]
```diff
@@ -190,21 +195,17 @@ protected IMvcRazorHost GetRazorHost()
             };
         }
 
-        private PrecompilationCacheEntry OnCacheMiss(ICacheSetContext cacheSetContext)
+        private CacheEntryOptions GetCacheEntryOptions(
+            RelativeFileInfo fileInfo,
+            PrecompilationCacheEntry cacheEntry)
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorPreCompiler.cs]: Do you need to pass `cacheEntry` in?

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs [hunk 8]
```diff
@@ -194,18 +196,17 @@ public class CompilerCache : ICompilerCache
             };
         }
 
-        private CompilerCacheEntry PopulateCacheSetContext(ICacheSetContext cacheSetContext)
+        private CacheEntryOptions GetCacheEntryOptions(string key, CompilerCacheEntry compilerCacheEntry)
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs]: I think all you need here is the relativePath. Could we simplify this method to accept the `RelativeFileInfo` and use that on L#202 and L#204?

**Comment by pranavkm** [comment]
:watch:

File: test/WebSites/MvcTagHelpersWebSite/Components/ProductsViewComponent.cs (hunk 3 above)
Comment (Tratcher) [inline test/WebSites/MvcTagHelpersWebSite/Components/ProductsViewComponent.cs]: It's designed to let you link two cache entries, not add triggers from arbitrary call sites. Don't bring it back without taking this up with @lodejard

File: src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs (hunk 1 above)
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs]: The name of this method is `GetOrAdd`

File: src/Microsoft.AspNet.Mvc.TagHelpers/CacheTagHelper.cs (hunk 4 above)
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.TagHelpers/CacheTagHelper.cs]: oops..good catch...

File: test/WebSites/MvcTagHelpersWebSite/Components/ProductsViewComponent.cs (hunk 3 above)
Comment (kichalla) [inline test/WebSites/MvcTagHelpersWebSite/Components/ProductsViewComponent.cs]: @pranavkm : I like your idea. 
@Tratcher: Sure, will check

**Comment by kichalla** [comment]
Updated. /cc: @Tratcher @pranavkm

File: samples/TagHelperSample.Web/Components/FeaturedMoviesComponent.cs [hunk 9]
```diff
@@ -11,31 +13,38 @@ namespace TagHelperSample.Web.Components
...
+            _cache = cache;
         }
 
         public IViewComponentResult Invoke()
         {
-            IExpirationTrigger trigger;
-            var movies = _moviesService.GetFeaturedMovies(out trigger);
-            
-            // Add custom triggers
-            EntryLinkHelpers.ContextLink.AddExpirationTriggers(new[] { trigger });
```
Comment (pranavkm) [inline samples/TagHelperSample.Web/Components/FeaturedMoviesComponent.cs]: cc @loudej \ @davidfowl 

This does take away the ability to add an arbitrary expiration trigger without caching it. I'm fine if this is the intended design for this, but it's different from what we'd initially discussed when designing the `CacheTagHelper`.

File: src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs [hunk 10]
```diff
@@ -43,22 +43,24 @@ public DefaultCodeTreeCache(IFileProvider fileProvider)
         public CodeTree GetOrAdd([NotNull] string pagePath,
                                  [NotNull] Func<IFileInfo, CodeTree> getCodeTree)
         {
-            return _codeTreeCache.GetOrSet(pagePath, getCodeTree, OnCacheMiss);
-        }
+            CodeTree codeTree;
+            if (!_codeTreeCache.TryGetValue(pagePath, out codeTree))
+            {
+                // GetOrAdd is invoked for each _ViewStart that might potentially exist in the path.
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs]: Unrelated, but could you change `_ViewStart` -> `_GlobalImport` in the comment?

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs [hunk 11]
```diff
@@ -194,18 +201,17 @@ public class CompilerCache : ICompilerCache
...
-        private CompilerCacheEntry PopulateCacheSetContext(ICacheSetContext cacheSetContext)
+        private MemoryCacheEntryOptions GetMemoryCacheEntryOptions(string normalizedPath, string relativePath)
         {
-            var entry = (CompilerCacheEntry)cacheSetContext.State;
-            cacheSetContext.AddExpirationTrigger(_fileProvider.Watch(entry.RelativePath));
+            var options = new MemoryCacheEntryOptions();
+            options.AddExpirationTrigger(_fileProvider.Watch(relativePath));
 
-            var globalImportPaths = ViewHierarchyUtility.GetGlobalImportLocations(cacheSetContext.Key);
+            var globalImportPaths = ViewHierarchyUtility.GetGlobalImportLocations(normalizedPath);
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs]: You should be ok passing in the `relativePath` here (in fact it's probably a bug in that we're using the `normalizedPath` here).

File: samples/TagHelperSample.Web/Components/FeaturedMoviesComponent.cs (hunk 9 above)
Comment (davidfowl) [inline samples/TagHelperSample.Web/Components/FeaturedMoviesComponent.cs]: It only does that for the top level. If you did a cache.Get within the tag helper, it would automagically pick up the link.

File: samples/TagHelperSample.Web/Views/Movies/Index.cshtml [hunk 12]
```diff
@@ -35,7 +35,7 @@
     </div>
     <div class="sidebar">
         <cache expires-after="TimeSpan.FromMinutes(20)">
-            @await Component.InvokeAsync("FeaturedMovies")
+            @Component.Invoke("FeaturedMovies")
```
Comment (pranavkm) [inline samples/TagHelperSample.Web/Views/Movies/Index.cshtml]: Why change this?

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs [hunk 13]
```diff
@@ -776,13 +743,18 @@ public void UpdateCacheContext_CopiesTriggersFromEntryLink()
...
+                    if(!cache.TryGetValue("key1", out tagHelperContent))
                     {
-                        new CancellationTokenTrigger(tokenSource.Token)
-                    });
-                    return Task.FromResult<TagHelperContent>(expectedContent);
+                        tagHelperContent = expectedContent;
+                        cache.Set(
+                            "key1",
+                            tagHelperContent,
+                            new MemoryCacheEntryOptions()
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/CacheTagHelperTest.cs]: Could you pull this out into a variable?

File: samples/TagHelperSample.Web/Components/FeaturedMoviesComponent.cs (hunk 9 above)
Comment (pranavkm) [inline samples/TagHelperSample.Web/Components/FeaturedMoviesComponent.cs]: Not sure I fully follow. With the previous design, I could add arbitrary expiration triggers to the tag helper without having to create new cache entries.

```
EntryLinkHelpers.ContentLink.AddExpirationTriggers(new[] { FileProvider.Watch("/user.dat") });
```

With this design you're forced to cache something for the linking to work. In this example, @kichalla is creating a `featured_movies` cache entry even though he doesn't actually particularly care about caching it.

**Comment by pranavkm** [comment]
:watch: for minor comments.

File: samples/TagHelperSample.Web/Views/Movies/Index.cshtml (hunk 12 above)
Comment (kichalla) [inline samples/TagHelperSample.Web/Views/Movies/Index.cshtml]: I changed it as the ViewComponent does not have a `InvokeAsync` method as such(yeah I know the view component has a fallback logic)..

**Comment by kichalla** [comment]
Updated. /cc: @pranavkm

File: src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs [hunk 14]
```diff
@@ -43,22 +43,24 @@ public DefaultCodeTreeCache(IFileProvider fileProvider)
         public CodeTree GetOrAdd([NotNull] string pagePath,
                                  [NotNull] Func<IFileInfo, CodeTree> getCodeTree)
         {
-            return _codeTreeCache.GetOrSet(pagePath, getCodeTree, OnCacheMiss);
-        }
+            CodeTree codeTree;
+            if (!_codeTreeCache.TryGetValue(pagePath, out codeTree))
+            {
+                // GetOrAdd is invoked for each _GlobalImport that might potentially exist in the path.
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs]: `_GlobalImport` is being renamed to `_ViewImport`. #2596

Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs]: yeah, right... @dougbu  mentioned it to me...i was planning to revert this change as his PR is going to have this change...

File: samples/TagHelperSample.Web/Components/FeaturedMoviesComponent.cs [hunk 15]
```diff
@@ -11,31 +13,38 @@ namespace TagHelperSample.Web.Components
...
+            var cacheKey = "featured_movies";
+            IEnumerable<FeaturedMovies> movies;
+            if (!_cache.TryGetValue(cacheKey, out movies))
+            {
+                IExpirationTrigger trigger;
+                movies = _moviesService.GetFeaturedMovies(out trigger);
+                _cache.Set(cacheKey, movies, new MemoryCacheEntryOptions().AddExpirationTrigger(trigger));
```
Comment (pranavkm) [inline samples/TagHelperSample.Web/Components/FeaturedMoviesComponent.cs]: Maybe make a comment this also affects the caching of the base thing. It's not particularly obvious looking at this.

**Comment by pranavkm** [comment]
:shipit:
